File: app/src/main/java/com/example/botcoin/za/co/botcoin/wallet/fragments/WithdrawFrag.java
package com.example.botcoin.za.co.botcoin.wallet.fragments;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.botcoin.R;
import com.example.botcoin.za.co.botcoin.utils.ConstantUtils;
import com.example.botcoin.za.co.botcoin.utils.GeneralUtils;
import com.example.botcoin.za.co.botcoin.utils.StringUtils;
import com.example.botcoin.za.co.botcoin.utils.WSCallUtilsCallBack;
import com.example.botcoin.za.co.botcoin.utils.WSCallsUtils;

import org.json.JSONObject;

import static android.content.Context.NOTIFICATION_SERVICE;

public class WithdrawFrag extends Fragment implements WSCallUtilsCallBack
{
    private final int REQ_CODE_WITHDRAW = 101;

    private EditText edTxtAmount;
    private EditText edTxtBeneficiaryId;
    private Button btnWithdraw;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.frag_withdraw, container, false);

        initUI(view);
        addWithdrawListener(view);

        return view;
    }

    private void initUI(View view)
    {
        this.edTxtAmount = view.findViewById(R.id.edTxtAmount);
        this.edTxtBeneficiaryId = view.findViewById(R.id.edTxtBeneficiaryId);
        this.btnWithdraw = view.findViewById(R.id.btnWithdraw);
    }

    private void addWithdrawListener(View view)
    {
        this.btnWithdraw = view.findViewById(R.id.btnWithdraw);
        this.btnWithdraw.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v)
            {
                if(!edTxtAmount.getText().toString().equals("") && !edTxtAmount.getText().toString().equals("0") && !edTxtBeneficiaryId.getText().toString().equals(""))
                {
                    if(GeneralUtils.isApiKeySet(getContext()))
                    {
                        withdrawal(edTxtAmount.getText().toString(), edTxtBeneficiaryId.getText().toString());
                    }else
                    {
                        GeneralUtils.createAlertDialog(getActivity(),"Luno API Credentials","Please set your Luno API credentials in order to use BotCoin!", false).show();
                    }
                }else
                {
                    GeneralUtils.createAlertDialog(getActivity(),"Withdrawal","Please provide an amount more than 0 and a valid beneficiary ID!", false).show();
                }
            }
        });
    }

    private void withdrawal(String amount, String beneficiaryId)
    {
        WSCallsUtils.post(this, REQ_CODE_WITHDRAW, StringUtils.GLOBAL_LUNO_URL + StringUtils.GLOBAL_ENDPOINT_WITHDRAWALS + GeneralUtils.buildWithdrawal(amount, beneficiaryId), "", GeneralUtils.getAuth(ConstantUtils.USER_KEY_ID, ConstantUtils.USER_SECRET_KEY));
    }

    @Override
    public void taskCompleted(String response, int reqCode)
    {
        if(response != null)
        {
            if(reqCode == REQ_CODE_WITHDRAW)
            {
                try
                {
                    JSONObject jsonObject = new JSONObject(response);

                    if(jsonObject != null)
                    {
                        notify("Withdrew " + this.edTxtAmount.getText().toString() + " Rands.", jsonObject.toString());
                    }

                }catch(Exception e)
                {
                    Log.e(ConstantUtils.BOTCOIN_TAG, "\nError: " + e.getMessage()
                            + "\nMethod: DonateFrag - taskCompleted"
                            + "\nRequest Code: " + reqCode
                            + "\nCreatedTime: " + GeneralUtils.getCurrentDateTime());
                }
            }
        }
    }

    public void notify(String title, String message)
    {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
        {
            Intent intent = new Intent();
            PendingIntent pendingIntent = PendingIntent.getActivity(getContext(), 0,intent,0);
            Notification notification = new Notification.Builder(getContext())
                    .setTicker(title)
                    .setContentTitle(title)
                    .setContentText(message)
                    .setSmallIcon(R.drawable.botcoin)
                    .addAction(R.drawable.luno_icon,"Action 1", pendingIntent)
                    .setContentIntent(pendingIntent).getNotification();

            notification.flags = Notification.FLAG_AUTO_CANCEL;

            NotificationManager notificationManager = (NotificationManager)getContext().getSystemService(NOTIFICATION_SERVICE);
            notificationManager.notify(0, notification);
        }

        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
        {
            Intent intent = new Intent();
            PendingIntent pendingIntent = PendingIntent.getActivity(getContext(), 0,intent,0);

            Notification notification  = new Notification.Builder(getContext())
                    .setContentTitle(title)
                    .setContentText(message)
                    .setSmallIcon(R.drawable.botcoin)
                    .setContentIntent(pendingIntent)
                    .setDefaults(Notification.DEFAULT_SOUND|Notification.DEFAULT_LIGHTS|Notification.DEFAULT_VIBRATE)
                    .setAutoCancel(true)
                    .build();

            NotificationManager notificationManager = (NotificationManager)getContext().getSystemService(NOTIFICATION_SERVICE);
            notificationManager.notify(0, notification);
        }

    }
}

File: app/src/main/java/za/co/botcoin/wallet/fragments/WithdrawFrag.kt
package za.co.botcoin.wallet.fragments

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.fragment.app.Fragment
import org.json.JSONObject
import za.co.botcoin.R
import za.co.botcoin.databinding.WithdrawFragmentBinding
import za.co.botcoin.utils.*
import za.co.botcoin.utils.GeneralUtils.buildWithdrawal
import za.co.botcoin.utils.GeneralUtils.createAlertDialog
import za.co.botcoin.utils.GeneralUtils.getAuth
import za.co.botcoin.utils.GeneralUtils.isApiKeySet

class WithdrawFrag : Fragment(R.layout.withdraw_fragment), WSCallUtilsCallBack {
    private lateinit var binding: WithdrawFragmentBinding
    private val REQ_CODE_WITHDRAW = 101

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        this.binding = WithdrawFragmentBinding.bind(view)

        addWithdrawListener(view)
    }

    private fun addWithdrawListener(view: View) {
        this.binding.btnWithdraw.setOnClickListener(View.OnClickListener {
            if (this.binding.edTxtAmount.text.toString() != "" && this.binding.edTxtAmount.text.toString() != "0" && this.binding.edTxtBeneficiaryId.text.toString() != "") {
                if (isApiKeySet(context)) {
                    withdrawal(this.binding.edTxtAmount.text.toString(), this.binding.edTxtBeneficiaryId.text.toString())
                } else {
                    createAlertDialog(activity, "Luno API Credentials", "Please set your Luno API credentials in order to use BotCoin!", false)!!.show()
                }
            } else {
                createAlertDialog(activity, "Withdrawal", "Please provide an amount more than 0 and a valid beneficiary ID!", false)!!.show()
            }
        })
    }

    private fun withdrawal(amount: String, beneficiaryId: String) {
        WSCallsUtils.post(this, REQ_CODE_WITHDRAW, StringUtils.GLOBAL_LUNO_URL + StringUtils.GLOBAL_ENDPOINT_WITHDRAWALS + buildWithdrawal(amount, beneficiaryId), "", getAuth(ConstantUtils.USER_KEY_ID!!, ConstantUtils.USER_SECRET_KEY!!))
    }

    override fun taskCompleted(response: String?, reqCode: Int) {
        if (response != null) {
            if (reqCode == REQ_CODE_WITHDRAW) {
                try {
                    val jsonObject = JSONObject(response)
                    if (jsonObject != null) {
                        notify("Withdrew " + this.binding.edTxtAmount.text.toString() + " Rands.", jsonObject.toString())
                    }
                } catch (e: Exception) {
                    Log.e(ConstantUtils.BOTCOIN_TAG, "Error: ${e.message} " +
                            "Method: DonateFrag - taskCompleted " +
                            "Request Code: $reqCode " +
                            "CreatedTime: ${GeneralUtils.getCurrentDateTime()}")
                }
            }
        }
    }

    private fun notify(title: String?, message: String?) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val intent = Intent()
            val pendingIntent = PendingIntent.getActivity(context, 0, intent, 0)
            val notification = Notification.Builder(context)
                    .setTicker(title)
                    .setContentTitle(title)
                    .setContentText(message)
                    .setSmallIcon(R.drawable.botcoin)
                    .addAction(R.drawable.luno_icon, "Action 1", pendingIntent)
                    .setContentIntent(pendingIntent).notification
            notification.flags = Notification.FLAG_AUTO_CANCEL
            val notificationManager = context!!.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.notify(0, notification)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val intent = Intent()
            val pendingIntent = PendingIntent.getActivity(context, 0, intent, 0)
            val notification = Notification.Builder(context)
                    .setContentTitle(title)
                    .setContentText(message)
                    .setSmallIcon(R.drawable.botcoin)
                    .setContentIntent(pendingIntent)
                    .setDefaults(Notification.DEFAULT_SOUND or Notification.DEFAULT_LIGHTS or Notification.DEFAULT_VIBRATE)
                    .setAutoCancel(true)
                    .build()
            val notificationManager = context!!.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.notify(0, notification)
        }
    }
}